D2M Pass 1: Attach Metal Layout

jdesousa-TT opened the pull request:
Add an initial pass to attach canonical `tt.metal_layout` to all tensors in the graph.

i.e.

```mlir
func.func @maximum(%arg0: tensor<64x128xf32>, %arg1: tensor<64x128xf32>) -> tensor<64x128xf32> {
  %0 = tensor.empty() : tensor<64x128xf32>
  %1 = "ttir.maximum"(%arg0, %arg1, %0) <{operandSegmentSizes = array<i32: 2, 1>}> : (tensor<64x128xf32>, tensor<64x128xf32>, tensor<64x128xf32>) -> tensor<64x128xf32>
  return %1 : tensor<64x128xf32>
}
```

becomes

```mlir
#l1_ = #tt.memory_space<l1>
#layout = #tt.metal_layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<64x128xf32, #l1_>>
module {
  func.func @maximum(%arg0: tensor<64x128xf32, #layout>, %arg1: tensor<64x128xf32, #layout>) -> tensor<64x128xf32, #layout> {
    %0 = tensor.empty() : tensor<64x128xf32, #layout>
    %1 = "ttir.maximum"(%arg0, %arg1, %0) <{operandSegmentSizes = array<i32: 2, 1>}> : (tensor<64x128xf32, #layout>, tensor<64x128xf32, #layout>, tensor<64x128xf32, #layout>) -> tensor<64x128xf32, #layout>
    return %1 : tensor<64x128xf32, #layout>
  }
}
```

The following remains the same because all tensors already have layout attributes:
```mlir
#l1_ = #tt.memory_space<l1>
#layout1 = #tt.metal_layout<(d0, d1) -> (d0, d1), undef, <4x4>, memref<64x96xf32, #l1_>>
#layout2 = #tt.metal_layout<(d0, d1) -> (d0, d1), undef, <4x1>, memref<64x32xf32, #l1_>>

func.func @reduceW(%arg0: tensor<256x384xf32, #layout1>) -> tensor<256x32xf32, #layout2> {
  %0 = tensor.empty() : tensor<256x32xf32, #layout2>
  %1 = "ttir.sum"(%arg0, %0) <{operandSegmentSizes = array<i32: 1, 1>,
                               dim_arg = [-1: i32],
                               keep_dim = true}> :
    (tensor<256x384xf32, #layout1>, tensor<256x32xf32, #layout2>) -> tensor<256x32xf32, #layout2>
  return %1 : tensor<256x32xf32, #layout2>
}

#layout3 = #tt.metal_layout<(d0, d1) -> (d0, d1), undef, <1x4>, memref<32x96xf32, #l1_>>
func.func @reduceH(%arg0: tensor<256x384xf32, #layout1>) -> tensor<32x384xf32, #layout3> {
  %0 = tensor.empty() : tensor<32x384xf32, #layout3>
  %1 = "ttir.sum"(%arg0, %0) <{operandSegmentSizes = array<i32: 1, 1>,
                               dim_arg = [-2: i32],
                               keep_dim = true}> :
    (tensor<256x384xf32, #layout1>, tensor<32x384xf32, #layout3>) -> tensor<32x384xf32, #layout3>
  return %1 : tensor<32x384xf32, #layout3>
}

#layout4 = #tt.metal_layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<32x32xf32, #l1_>>
func.func @reduceWH(%arg0: tensor<256x384xf32, #layout1>) -> tensor<32x32xf32, #layout4> {
  %0 = tensor.empty() : tensor<32x32xf32, #layout4>
  %1 = "ttir.sum"(%arg0, %0) <{operandSegmentSizes = array<i32: 1, 1>,
                               dim_arg = [-1: i32, -2: i32],
                               keep_dim = true}> :
    (tensor<256x384xf32, #layout1>, tensor<32x32xf32, #layout4>) -> tensor<32x32xf32, #layout4>
  return %1 : tensor<32x32xf32, #layout4>
}

func.func @maxReduceWH(%arg0: tensor<256x384xf32, #layout1>) -> tensor<32x32xf32, #layout4> {
  %0 = tensor.empty() : tensor<32x32xf32, #layout4>
  %1 = "ttir.max" (%arg0, %0) <{operandSegmentSizes = array<i32: 1, 1>,
                               dim_arg = [-1: i32, -2: i32],
                               keep_dim = true}> :
    (tensor<256x384xf32, #layout1>, tensor<32x32xf32, #layout4>) -> tensor<32x32xf32, #layout4>
  return %1 : tensor<32x32xf32, #layout4>
}
```

nsmithtt reviewed (commented):
Thanks @jdesousa-TT for getting the ball rolling on this!

A few comments/questions:
- I don't see how this works for values that don't participate in non-FuncOp operands and results.  I.e. a series of TTIR ops.
- I think the "MLIR way" of doing this is to use a TypeConverter, see ` lib/Dialect/TTIR/Transforms/Layout.cpp` which already does this.  I think the `TTIRLayout` might have everything that we're looking for, Jovan refactored a few things here so I'm not 100% sure what the state of the pass is today, but we should try running it in isolation to see if it just works or what tweaks we could make.
- We need to reorder this pass to happen after TTIRImplicitDevice in TTMetalBackenPipeline.
- We need a simple test

jdesousa-TT commented:
> Thanks @jdesousa-TT for getting the ball rolling on this!
> 
> A few comments/questions:
> 
> * I don't see how this works for values that don't participate in non-FuncOp operands and results.  I.e. a series of TTIR ops.
> * I think the "MLIR way" of doing this is to use a TypeConverter, see ` lib/Dialect/TTIR/Transforms/Layout.cpp` which already does this.  I think the `TTIRLayout` might have everything that we're looking for, Jovan refactored a few things here so I'm not 100% sure what the state of the pass is today, but we should try running it in isolation to see if it just works or what tweaks we could make.
> * We need to reorder this pass to happen after TTIRImplicitDevice in TTMetalBackenPipeline.
> * We need a simple test

Hey Nick,

- I was working under the assumption all of our IR (including TTIR ops) are nested within function(s).
- I will investigate this further
- I did not include this in the pipeline yet, as it might interfere with the existing layout pass. I figured we would build out this new milestone and refactor the pipeline once the individual passes are running and tested
- I can add a couple simple tests for this!

Let me know what you think, thanks!

nsmithtt commented:
- I was working under the assumption all of our IR (including TTIR ops) are nested within function(s). Is this not the case? I just haven't found an existing test, but maybe I'm missing some key info :)

It's possible I'm missing something, but it seems like this case:
```mlir
func.func @forward(%arg0: tensor<1x784xf32>, %arg1: tensor<784x512xf32>, %arg2: tensor<512xf32>) -> tensor<1x10xf32> {                                                                                 
  %0 = tensor.empty() : tensor<1x512xf32>                                                                                  
  %1 = "ttir.matmul"(%arg0, %arg1, %0) : (tensor<1x784xf32>, tensor<784x512xf32>, tensor<1x512xf32>) -> tensor<1x512xf32>      
  %2 = tensor.empty() : tensor<1x512xf32>                                                                                                                   
  %3 = "ttir.add"(%1, %arg2, %2) <{operandSegmentSizes = array<i32: 2, 1>}> : (tensor<1x512xf32>, tensor<512xf32>, tensor<1x512xf32>) -> tensor<1x512xf32>      
  return %3 : tensor<1x10xf32>      
}
```

It wasn't clear to me at first pass how values `%1` and `%2` are updated.

jdesousa-TT commented:
Ah, I see. I match the function ops and walk their contents, searching for any operands and/or results that contain tensor types:
```
op.walk([&](Operation *op) {
      if (isa<func::FuncOp>(op)) {
        auto funcOp = mlir::cast<func::FuncOp>(op);
        std::vector inputs = std::vector<Type>();
        std::vector results = std::vector<Type>();
        for (auto operand : funcOp.getFunctionType().getInputs()) {
          if (isa<RankedTensorType>(operand)) {
            inputs.push_back(appendLayoutToType(operand));
          }
        }
        for (auto result : funcOp.getFunctionType().getResults()) {
          if (isa<RankedTensorType>(result)) {
            results.push_back(appendLayoutToType(result));
          }
        }
        mlir::FunctionType newFuncType =
            FunctionType::get(funcOp.getContext(), inputs, results);
        funcOp.setFunctionType(newFuncType);
        return;
      }
      for (auto operand : op->getOperands()) {
        appendLayoutToValue(operand);
      }
      for (auto result : op->getResults()) {
        appendLayoutToValue(result);
      }
    });
```

jdesousa-TT commented:
>  I think the TTIRLayout might have everything that we're looking for, Jovan refactored a few things here so I'm not 100% sure what the state of the pass is today, but we should try running it in isolation to see if it just works or what tweaks we could make.

Running this pass does what we are looking for, but some extra as well. Do we want this attach metal layout pass to include generating to_layout ops for tiling tensors for input to other ops? Or exclusively adding a metal layout attribute to those tensors without one? Here's an example of running the existing layout pass:

Before:
```mlir
func.func @maximum(%arg0: tensor<64x128xf32>, %arg1: tensor<64x128xf32>) -> tensor<64x128xf32> {
  // CHECK: %[[C:.*]] = "ttmetal.alloc"[[C:.*]]
  %0 = tensor.empty() : tensor<64x128xf32>
  // CHECK: %[[C:.*]] = "ttmetal.dispatch"[[C:.*]]
  %1 = "ttir.maximum"(%arg0, %arg1, %0) <{operandSegmentSizes = array<i32: 2, 1>}> : (tensor<64x128xf32>, tensor<64x128xf32>, tensor<64x128xf32>) -> tensor<64x128xf32>
  return %1 : tensor<64x128xf32>
}
```

After:
```mlir
#device = #tt.device<workerGrid = #tt.grid<8x8, (d0, d1) -> (0, d0, d1)>, l1Map = (d0, d1)[s0, s1] -> (0, d0 floordiv s0, d1 floordiv s1, (d0 mod s0) * s1 + d1 mod s1), dramMap = (d0, d1)[s0, s1] -> (0, 0, ((((d0 floordiv s0) * 8 + d1 floordiv s1) * (s1 * s0) + (d0 mod s0) * s1 + d1 mod s1) floordiv 8192) mod 12, (((d0 floordiv s0) * 8 + d1 floordiv s1) * (s1 * s0) + (d0 mod s0) * s1 + d1 mod s1) floordiv 98304 + (((d0 floordiv s0) * 8 + d1 floordiv s1) * (s1 * s0) + (d0 mod s0) * s1 + d1 mod s1) mod 8192), meshShape = , chipIds = [0]>
#dram = #tt.memory_space<dram>
#system = #tt.memory_space<system>
#system_desc = #tt.system_desc<[{role = host, target_triple = "x86_64-pc-linux-gnu"}], [{arch = <wormhole_b0>, grid = 8x8, l1_size = 1499136, num_dram_channels = 12, dram_channel_size = 1073741824, noc_l1_address_align_bytes = 16, pcie_address_align_bytes = 32, noc_dram_address_align_bytes = 32, l1_unreserved_base = 1024, erisc_l1_unreserved_base = 1024, dram_unreserved_base = 1024, dram_unreserved_end = 1073741824, physical_cores = {worker = [ 0x0,  0x1,  0x2,  0x3,  0x4,  0x5,  0x6,  0x7,  1x0,  1x1,  1x2,  1x3,  1x4,  1x5,  1x6,  1x7,  2x0,  2x1,  2x2,  2x3,  2x4,  2x5,  2x6,  2x7,  3x0,  3x1,  3x2,  3x3,  3x4,  3x5,  3x6,  3x7,  4x0,  4x1,  4x2,  4x3,  4x4,  4x5,  4x6,  4x7,  5x0,  5x1,  5x2,  5x3,  5x4,  5x5,  5x6,  5x7,  6x0,  6x1,  6x2,  6x3,  6x4,  6x5,  6x6,  6x7,  7x0,  7x1,  7x2,  7x3,  7x4,  7x5,  7x6,  7x7] dram = [ 8x0,  9x0,  10x0,  8x1,  9x1,  10x1,  8x2,  9x2,  10x2,  8x3,  9x3,  10x3]}, supported_data_types = [<f32>, <f16>, <bf16>, <bfp_f8>, <bfp_bf8>, <bfp_f4>, <bfp_bf4>, <bfp_f2>, <bfp_bf2>, <u32>, <u16>, <u8>], supported_tile_sizes = [ 4x16,  16x16,  32x16,  4x32,  16x32,  32x32], num_cbs = 32}], [0], [3 : i32], [ 0x0x0x0]>
#layout = #tt.metal_layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<64x128xf32, #system>>
#layout1 = #tt.metal_layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<2x4x!tt.tile<32x32, f32>, #dram>, interleaved>
module attributes {tt.device = #device, tt.system_desc = #system_desc} {
  func.func @maximum(%arg0: tensor<64x128xf32, #layout>, %arg1: tensor<64x128xf32, #layout>) -> tensor<64x128xf32, #layout> {
    %0 = tensor.empty() : tensor<64x128xf32, #layout1>
    %1 = "ttir.to_layout"(%arg0, %0) : (tensor<64x128xf32, #layout>, tensor<64x128xf32, #layout1>) -> tensor<64x128xf32, #layout1>
    %2 = tensor.empty() : tensor<64x128xf32, #layout1>
    %3 = "ttir.to_layout"(%arg1, %2) : (tensor<64x128xf32, #layout>, tensor<64x128xf32, #layout1>) -> tensor<64x128xf32, #layout1>
    %4 = tensor.empty() : tensor<64x128xf32, #layout1>
    %5 = "ttir.maximum"(%1, %3, %4) <{operandSegmentSizes = array<i32: 2, 1>}> : (tensor<64x128xf32, #layout1>, tensor<64x128xf32, #layout1>, tensor<64x128xf32, #layout1>) -> tensor<64x128xf32, #layout1>
    %6 = tensor.empty() : tensor<64x128xf32, #layout>
    %7 = "ttir.to_layout"(%5, %6) : (tensor<64x128xf32, #layout1>, tensor<64x128xf32, #layout>) -> tensor<64x128xf32, #layout>
    return %7 : tensor<64x128xf32, #layout>
  }
}
```

jdesousa-TT commented:
@nsmithtt Please take another look at this PR. I changed over to using a TypeConverter and RewritePattern (heavily borrowed from TTIRLayout) which should be more MLIR-esque. I've added a test as well for this pass. Please verify that this is the correct location for the test.

nsmithtt reviewed (approved):
Looks good to me!
